Maple2.Server.Game/PacketHandlers/MasteryHandler.cs

```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Maple2.Database.Storage;
+using Maple2.Model.Error;
+using Maple2.Model.Game;
+using Maple2.Model.Metadata;
+using Maple2.PacketLib.Tools;
+using Maple2.Server.Core.Constants;
+using Maple2.Server.Core.PacketHandlers;
+using Maple2.Server.Game.Packets;
+using Maple2.Server.Game.Session;
+
+namespace Maple2.Server.Game.PacketHandlers;
+
+public class MasteryHandler : PacketHandler<GameSession> {
+    public override RecvOp OpCode => RecvOp.Mastery;
+
+    private enum Command : byte {
+        Reward = 1,
+        Craft = 2,
+    }
+
+    #region Autofac Autowired
+    // ReSharper disable MemberCanBePrivate.Global
+    public ItemMetadataStorage ItemMetadata { get; init; } = null!;
+    public TableMetadataStorage TableMetadata { private get; init; } = null!;
+    // ReSharper restore All
+    #endregion
+
+    public override void Handle(GameSession session, IByteReader packet) {
+        var command = packet.Read<Command>();
+        switch (command) {
+            case Command.Craft:
+                HandleCraft(session, packet);
+                break;
+        }
+    }
+
+    private void HandleCraft(GameSession session, IByteReader packet) {
+        int recipeId = packet.ReadInt();
+
+        if (!TableMetadata.MasteryRecipeTable.Entries.TryGetValue(recipeId, out MasteryRecipeTable.Entry? entry)) {
+            return;
+        }
+
+        // TODO: Check if player has completed the required quests.
+
+        if (session.Mastery[entry.Type] < entry.RequiredMastery) {
+            session.Send(MasteryPacket.Error(MasteryError.s_mastery_error_lack_mastery));
+        }
+
+        if (session.Currency.CanAddMeso(-entry.RequiredMeso) != -entry.RequiredMeso) {
+            session.Send(MasteryPacket.Error(MasteryError.s_mastery_error_lack_meso));
+            return;
+        }
+
+        Dictionary<int, IEnumerable<Item>> itemDictionary = new Dictionary<int, IEnumerable<Item>>();
```

Comment:
it's okay to use `var` here since the type is obvious due to the constructor being called.